File: Tutorials/Codes/Advanced_Robotics_Course/Chapter11/Lesson2/Chapter11_Lesson2.cpp
#include <iostream>
#include <vector>
#include <Eigen/Dense>

using State = Eigen::Vector4d;   // [q1, q2, dq1, dq2]
using Action = Eigen::Vector2d;  // [tau1, tau2]

struct Sample {
    State s;
    Action a;
};

class LinearPolicy {
public:
    Eigen::Matrix<double, 2, 4> W;
    Action b;

    LinearPolicy() {
        W.setZero();
        b.setZero();
    }

    Action operator()(const State &s) const {
        return W * s + b;
    }
};

// Dummy expert PD controller
Action expertPolicy(const State &s) {
    Eigen::Vector2d q_des(0.5, -0.5);
    Eigen::Vector2d q = s.segment<2>(0);
    Eigen::Vector2d dq = s.segment<2>(2);
    double Kp = 5.0, Kd = 1.0;
    return Kp * (q_des - q) - Kd * dq;
}

// Simple discrete-time dynamics for illustration
State simulate(const State &s, const Action &a, double dt = 0.02) {
    Eigen::Vector2d q = s.segment<2>(0);
    Eigen::Vector2d dq = s.segment<2>(2);
    Eigen::Vector2d ddq = a; // placeholder
    Eigen::Vector2d dq_next = dq + ddq * dt;
    Eigen::Vector2d q_next = q + dq_next * dt;
    State s_next;
    s_next << q_next, dq_next;
    return s_next;
}

// Least-squares fit of linear policy: minimize sum ||W s + b - a||^2
void trainBC(LinearPolicy &pi, const std::vector<Sample> &data, int iters = 100, double lr = 1e-3) {
    for (int k = 0; k < iters; ++k) {
        Eigen::Matrix<double, 2, 4> gradW = Eigen::Matrix<double, 2, 4>::Zero();
        Action gradb = Action::Zero();
        for (const auto &sample : data) {
            Action pred = pi(sample.s);
            Action err = pred - sample.a;
            gradW += err * sample.s.transpose();
            gradb += err;
        }
        gradW /= static_cast<double>(data.size());
        gradb /= static_cast<double>(data.size());
        pi.W -= lr * gradW;
        pi.b -= lr * gradb;
    }
}

std::vector<Sample> collectExpertRollout(int T) {
    std::vector<Sample> traj;
    State s = State::Zero();
    for (int t = 0; t < T; ++t) {
        Action a = expertPolicy(s);
        traj.push_back({s, a});
        s = simulate(s, a);
    }
    return traj;
}

// DAgger training loop
void daggerTrain(int numIters = 5, int T = 200) {
    LinearPolicy pi;
    std::vector<Sample> dataset = collectExpertRollout(T);

    for (int i = 0; i < numIters; ++i) {
        trainBC(pi, dataset, 200, 1e-3);

        // Rollout current policy
        State s = State::Zero();
        std::vector<Sample> newSamples;
        for (int t = 0; t < T; ++t) {
            Action a_pi = pi(s);
            // Query expert at visited state s
            Action a_exp = expertPolicy(s);
            newSamples.push_back({s, a_exp});
            s = simulate(s, a_pi);
        }

        // Aggregate
        dataset.insert(dataset.end(), newSamples.begin(), newSamples.end());
        std::cout << "DAgger iter " << (i+1)
                  << ", dataset size = " << dataset.size() << std::endl;
    }
}

int main() {
    daggerTrain();
    return 0;
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter11/Lesson2/Chapter11_Lesson2.java
import java.util.ArrayList;
import java.util.List;

class Sample {
    double[] s; // length 4
    double[] a; // length 2
    Sample(double[] s, double[] a) {
        this.s = s;
        this.a = a;
    }
}

class LinearPolicy {
    // a = W s + b, W: 2x4, b: 2
    double[][] W = new double[2][4];
    double[] b = new double[2];

    double[] act(double[] s) {
        double[] a = new double[2];
        for (int i = 0; i < 2; i++) {
            a[i] = b[i];
            for (int j = 0; j < 4; j++) {
                a[i] += W[i][j] * s[j];
            }
        }
        return a;
    }
}

public class DAggerDemo {

    static double[] expertPolicy(double[] s) {
        double[] q_des = {0.5, -0.5};
        double[] q = {s[0], s[1]};
        double[] dq = {s[2], s[3]};
        double Kp = 5.0, Kd = 1.0;
        double[] tau = new double[2];
        for (int i = 0; i < 2; i++) {
            tau[i] = Kp * (q_des[i] - q[i]) - Kd * dq[i];
        }
        return tau;
    }

    static double[] simulate(double[] s, double[] a, double dt) {
        double q1 = s[0], q2 = s[1];
        double dq1 = s[2], dq2 = s[3];
        double ddq1 = a[0], ddq2 = a[1];
        double dq1n = dq1 + ddq1 * dt;
        double dq2n = dq2 + ddq2 * dt;
        double q1n = q1 + dq1n * dt;
        double q2n = q2 + dq2n * dt;
        return new double[]{q1n, q2n, dq1n, dq2n};
    }

    static List<Sample> collectExpertRollout(int T) {
        List<Sample> traj = new ArrayList<>();
        double[] s = {0, 0, 0, 0};
        for (int t = 0; t < T; t++) {
            double[] a = expertPolicy(s);
            traj.add(new Sample(s.clone(), a.clone()));
            s = simulate(s, a, 0.02);
        }
        return traj;
    }

    static void trainBC(LinearPolicy pi, List<Sample> data, int iters, double lr) {
        for (int it = 0; it < iters; it++) {
            double[][] gradW = new double[2][4];
            double[] gradb = new double[2];
            for (Sample sample : data) {
                double[] pred = pi.act(sample.s);
                double[] err = {pred[0] - sample.a[0], pred[1] - sample.a[1]};
                for (int i = 0; i < 2; i++) {
                    gradb[i] += err[i];
                    for (int j = 0; j < 4; j++) {
                        gradW[i][j] += err[i] * sample.s[j];
                    }
                }
            }
            int N = data.size();
            for (int i = 0; i < 2; i++) {
                gradb[i] /= N;
                for (int j = 0; j < 4; j++) gradW[i][j] /= N;
            }
            for (int i = 0; i < 2; i++) {
                pi.b[i] -= lr * gradb[i];
                for (int j = 0; j < 4; j++) {
                    pi.W[i][j] -= lr * gradW[i][j];
                }
            }
        }
    }

    static void daggerTrain() {
        int numIters = 5;
        int T = 200;
        LinearPolicy pi = new LinearPolicy();
        List<Sample> dataset = collectExpertRollout(T);

        for (int i = 0; i < numIters; i++) {
            trainBC(pi, dataset, 200, 1e-3);
            List<Sample> newSamples = new ArrayList<>();
            double[] s = {0, 0, 0, 0};
            for (int t = 0; t < T; t++) {
                double[] a_pi = pi.act(s);
                double[] a_exp = expertPolicy(s);
                newSamples.add(new Sample(s.clone(), a_exp.clone()));
                s = simulate(s, a_pi, 0.02);
            }
            dataset.addAll(newSamples);
            System.out.println("DAgger iter " + (i + 1) + ", dataset size = " + dataset.size());
        }
    }

    public static void main(String[] args) {
        daggerTrain();
    }
}
      
